### PR TITLE
support custom raft log to boost batch insert performance

### PR DESCRIPTION
**Pull request by coocood**
The custom raft log format is optional, we can still build TiKV compatible raft log.

After this optimization, the batch insert performance is doubled, exceeds the latest tikv.

Tested on a 40 core linux server

```
create table t (id bigint primary key auto_increment, c bigint)
```

Then run 8 workers do

```
for {
     insert t (c) values ... // 500 rows
}
```

TiKV:
```
177000
185500
184500
178500
179000
181000
176000
182500
179500
177000
178000
178500
173500
177000
183000
176000
177500
179000
178500
185000
184000
183500
```

Unistore:
```
255000
227000
227000
233500
234500
255500
247000
209000
252500
257000
254500
256500
214500
264500
243000
256000
257500
```

### CODE REVIEW

**Comment by cfzjywxk** [comment]
Add some documents of the design and optimization points about the new log format?   :eyes:

**Comment by coocood** [comment]
@cfzjywxk PTAL

File: tikv/raftstore/raftlog/custom.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	TypeRolback             CustomRaftLogType = 3
+	TypePessimisticLock     CustomRaftLogType = 4
+	TypePessimisticRollback CustomRaftLogType = 5
+)
+
+// CustomRaftLog is the raft log format for unistore to store Prewrite/Commit/PessimisticLock.
+//  | flag(1) | type(1) | version(2) | header(40) | entries
+//
+// It reduces the cost of marshal/unmarshal and avoid DB lookup during apply.
+type CustomRaftLog struct {
```
Comment (cfzjywxk) [inline tikv/raftstore/raftlog/custom.go]: Should we add some bench cases for the log generation and serialization

File: tikv/mvcc.go [hunk 2]
```diff
@@ -182,7 +182,6 @@ func (store *MVCCStore) PessimisticLock(reqCtx *requestCtx, req *kvrpcpb.Pessimi
 				return nil, nil
 			}
 			// Single statement rollback key, we can overwrite it.
-			batch.PessimisticRollback(m.Key)
```
[the remark this answers is not in the document]
Comment (coocood) [reply tikv/mvcc.go]: Yes

File: tikv/raftstore/raftlog/custom.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+	header *CustomHeader
+	Data   []byte
+}
+
+func NewCustom(data []byte) *CustomRaftLog {
```
[the remark this answers is not in the document]
Comment (coocood) [reply tikv/raftstore/raftlog/custom.go]: Yes

File: tikv/raftstore/raftlog/custom.go (hunk 1 above)
Comment (coocood) [reply tikv/raftstore/raftlog/custom.go]: It is all plain append, the bench result is not very helpful.

File: tikv/raftstore/applier.go [hunk 4]
```diff
@@ -434,7 +433,11 @@ func notifyStaleReq(term uint64, cb *Callback) {
 }
 
 /// Checks if a write is needed to be issued before handling the command.
-func shouldWriteToEngine(cmd *raft_cmdpb.RaftCmdRequest, wbKeys int) bool {
+func shouldWriteToEngine(rlog raftlog.RaftLog, wbKeys int) bool {
+	cmd := rlog.GetRaftCmdRequest()
+	if cmd == nil {
+		return false
```
Comment (bobotu) [inline tikv/raftstore/applier.go]: Is checking wbKeys >= WriteBatchMaxKeys not required when using new raft log?

File: tikv/raftstore/applier.go [hunk 5]
```diff
@@ -636,15 +639,22 @@ func (a *applier) handleRaftEntryNormal(aCtx *applyContext, entry *eraftpb.Entry
 	index := entry.Index
 	term := entry.Term
 	if len(entry.Data) > 0 {
-		cmd := new(raft_cmdpb.RaftCmdRequest)
-		err := cmd.Unmarshal(entry.Data)
-		if err != nil {
-			panic(err)
+		var rlog raftlog.RaftLog
+		if entry.Data[0] == raftlog.CustomRaftLogFlag {
```
Comment (bobotu) [inline tikv/raftstore/applier.go]: Is it possible that the first byte of `raft_cmdpb.RaftCmdRequest` equals to `raftlog.CustomRaftLogFlag`?

File: tikv/raftstore/peer_storage.go [hunk 6]
```diff
@@ -614,7 +614,7 @@ func (ps *PeerStorage) Snapshot() (eraftpb.Snapshot, error) {
 	return snap, raft.ErrSnapshotTemporarilyUnavailable
 }
 
-// Append the given entries to the raft log using previous last index or self.last_index.
+// AppendLock the given entries to the raft log using previous last index or self.last_index.
```
Comment (bobotu) [inline tikv/raftstore/peer_storage.go]: Unrelated change.

File: tikv/raftstore/applier.go (hunk 5 above)
Comment (cfzjywxk) [reply tikv/raftstore/applier.go]: 👀

Comment (coocood) [reply tikv/raftstore/applier.go]: No, the first byte of the PB is always the same.

File: tikv/raftstore/applier.go (hunk 4 above)
Comment (coocood) [reply tikv/raftstore/applier.go]: The check is not required, it's based on benchmarks of RocksDB.
We probably gonna remove it totally.

**Comment by cfzjywxk** [comment]
LGTM,   `coocood` codes fast like thunder :fast_forward:

**Comment by coocood** [comment]
@bobotu PTAL
